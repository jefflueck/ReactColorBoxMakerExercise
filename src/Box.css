.Box {
  outline: 3px solid black;
}

.Box-Button {
  text-align: center;
  width: 40px;
  margin-top: 10px;
  margin-bottom: 30px;
}
