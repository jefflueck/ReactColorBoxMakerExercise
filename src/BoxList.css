.BoxList {
  margin: 0;
  padding: 0;
}

.NewBoxForm {
  text-align: center;
  display: block;
}

.Boxes {
  display: flex;
  justify-content: space-around;
  flex-wrap: wrap;
  margin-top: 20px;
  margin-bottom: 20px;
}
