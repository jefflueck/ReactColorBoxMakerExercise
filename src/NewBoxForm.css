label {
  text-align: center;
  display: block;
  margin: 10px 0px;
}

button {
  display: block;
  margin: 0 auto;
  margin-top: 20px;
}
