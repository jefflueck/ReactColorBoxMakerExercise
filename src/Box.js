import React from 'react';
import './Box.css';

const Box = ({ id, backgroundColor, width, height, handleRemove }) => {
  const remove = () => handleRemove(id);
  return (
    <div>
      <div
        className="Box"
        style={{
          id: { id },
          handleRemove,
          backgroundColor,
          width: `${width}px`,
          height: `${height}px`,
        }}
      ></div>
      <button className="Box-Button" onClick={remove}>
        ❌
      </button>
    </div>
  );
};

export default Box;
