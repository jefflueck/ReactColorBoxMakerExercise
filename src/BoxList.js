import React, { useState } from 'react';
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import { v4 as uuidv4 } from 'uuid';
import './BoxList.css';

const BoxList = () => {
  const initialState = [];
  const [box, setBoxes] = useState(initialState);
  const addBox = (newBox) => {
    setBoxes((box) => [...box, { ...newBox, id: uuidv4() }]);
    console.log(newBox);
  };
  // solution
  const remove = (id) => {
    setBoxes((boxes) => boxes.filter((box) => box.id !== id));
  };

  return (
    <div className="BoxList">
      <div className="NewBoxForm">
        <NewBoxForm addBox={addBox} />
      </div>
      <div className="Boxes">
        {box.map(({ id, backgroundColor, width, height }) => (
          <Box
            id={id}
            handleRemove={remove}
            backgroundColor={backgroundColor}
            width={width}
            height={height}
            key={id}
          />
        ))}
      </div>
    </div>
  );
};

export default BoxList;
