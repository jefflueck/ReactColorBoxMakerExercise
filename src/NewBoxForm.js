import React, { useState } from 'react';
import './NewBoxForm.css';

const NewBoxForm = ({ addBox }) => {
  const initialState = {
    backgroundColor: '',
    width: '',
    height: '',
  };
  const [formData, setFormData] = useState(initialState);
  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((formData) => ({
      ...formData,
      [name]: value,
    }));
  };
  const handleSubmit = (e) => {
    e.preventDefault();
    addBox({ ...formData });
    setFormData(initialState);
  };
  return (
    <>
      <form onSubmit={handleSubmit}>
        <label htmlFor="backgroundColor">Color: </label>
        <input
          id="backgroundColor"
          type="text"
          name="backgroundColor"
          placeholder="New Box Color"
          value={formData.backgroundColor}
          onChange={handleChange}
        />
        <label htmlFor="width">Width: </label>
        <input
          id="width"
          type="number"
          name="width"
          placeholder="New Box Width"
          value={formData.width}
          onChange={handleChange}
        />
        <label htmlFor="height">Height: </label>
        <input
          id="height"
          type="number"
          name="height"
          placeholder="New Box height"
          value={formData.height}
          onChange={handleChange}
        />
        <button>Add Box</button>
      </form>
    </>
  );
};

export default NewBoxForm;
