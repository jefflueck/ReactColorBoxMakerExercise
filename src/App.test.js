import { render, fireEvent } from '@testing-library/react';
import App from './App';
import BoxList from './BoxList';
import NewBoxForm from './NewBoxForm';
import Box from './Box';

test('it renders the App', () => {
  render(<App />);
});

test('it renders a Box List Component', () => {
  render(<BoxList />);
  render(<NewBoxForm />);
});

test('it matches snapshot for Box in BoxList', () => {
  const { asFragment } = render(
    <Box background-color="red" width="200px" height="200px" />
  );
  expect(asFragment()).toMatchSnapshot();
});

test('it should add a new box', () => {
  const { queryByText, getByLabelText } = render(<BoxList />);
  const input = getByLabelText('Color:');
  const width = getByLabelText('Width:');
  const height = getByLabelText('Height:');
  const btn = queryByText('Add Box');
  fireEvent.change(input, width, height, {
    target: { backgroundColor: 'red', width: '200px', height: '200px' },
  });
  fireEvent.click(btn);
  expect(<Box background-color={'red'} width={'200px'} height={'200px'} />);
});
