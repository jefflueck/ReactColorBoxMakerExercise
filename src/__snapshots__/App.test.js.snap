// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot for Box in BoxList 1`] = `
<DocumentFragment>
  <div>
    <div
      class="Box"
    />
    <button
      class="Box-Button"
    >
      ‚ùå
    </button>
  </div>
</DocumentFragment>
`;
